### PR TITLE
Listen on loopback address during tests

### PR DESCRIPTION
**Pull request by casey**
This is a bit weird, since there's no good way to test that we're
actually listening on 0.0.0.0.

### CODE REVIEW

**Review by soenkehahn — APPROVED** [review]
I would've thought we would pass in e.g. `--address 127.0.0.1` during tests. (And implement `--address` first.) But I think this is also fine, since we do have a smoke test.

**Comment by casey** [comment]
That sounds fine to me. I implemented `--address`. Looking up an address is somewhat complicated, because `ToSocketAddrs` does host resolution, and so might produce an I/O error. Also, the OS might return non-IP addresses, which are ignored, so the resulting iterator might be empty, which also needs to be handled.

I added a test that checks that the right error is produced with the right hostname.

File: src/server.rs [hunk 1]
```diff
@@ -43,7 +53,7 @@ mod tests {
   use std::net::IpAddr;
 
   #[test]
-  fn listen_on_all_local_ip_addresses() {
+  fn listen_on_localhost() {
```
Comment (soenkehahn) [inline src/server.rs:56]:
```suggestion
  fn listen_on_localhost_by_default_in_tests() {
```

File: tests/smoke.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@ use std::{
 };
 
 #[test]
-fn server_listens_on_localhost_8080() {
+fn server_listens_on_all_ip_addresses_port_8080() {
```
Comment (soenkehahn) [inline tests/smoke.rs:10]:
```suggestion
fn server_listens_on_all_ip_addresses_on_port_8080() {
```

**Review by soenkehahn — APPROVED** [review]

**Pull request merged**